zilliqa/src/consensus.rs

```diff
@@ -472,87 +548,111 @@ impl Consensus {
-                        tx.hash(),
-                        tx.transaction.nonce,
-                        self.state.must_get_account(tx.from_addr).nonce,
-                        retries,
+                // if it is less than the iter nonce, we can skip it
+                // this just means there are 'priority txs' with the same nonce but different fees
+                if txn.nonce < iter_nonce {
+                    trace!("lower priority tx ignored {} vs {}", txn.nonce, iter_nonce);
+                    continue;
+                }
+
+                // If the nonce is too high, we will take another pass
+                if txn.nonce > iter_nonce {
+                    trace!(
+                        "TX nonce {} too high, iter nonce {}, retry {} of {}",
+                        txn.nonce,
+                        iter_nonce,
+                        txn.retries,
                         self.config.tx_retries
                     );
 
-                if *retries >= self.config.tx_retries {
-                    warn!(?tx, "Tranaction has exceeded retries and will been removed");
-                    tx_hashes_invalid.push(tx.hash());
-                    continue;
+                    if txn.retries >= self.config.tx_retries {
+                        warn!(?txn, "Tranaction has exceeded retries and all pending from this account will been removed");
+                        remove_all_txs = true;
+                    }
+
+                    // all other TXs will have a higher nonce, so we can break
+                    break;
                 }
-            }
 
-            // Tx nonce is good to add, we need to check that there isn't already a tx with the same
-            // from address, though (no need to drop though)
-            if tx_from_addrs.contains(&tx.from_addr) {
-                warn!(
-                    "Transaction from address {} already exists in block and will not be added",
-                    tx.from_addr
-                );
-                continue;
+                // tx nonce == iter nonce, so we can add it
+                match self.new_transactions.get(&txn.hash) {
+                    Some(tx) => {
+                        trace!("Adding tx to execute: {:?}", tx);
+                        ret.push(tx.clone());
+                        iter_nonce += 1;
+                    }
+                    None => {
+                        warn!(
+                            "Transaction {} not found in mempool, but was in priority txs",
+                            txn.hash
+                        );
+                        remove_all_txs = true;
+                    }
+                }
             }
 
-            // Tx appears to be good, add it to the list
-            tx_from_addrs.insert(tx.from_addr);
-            ret.push(tx.clone());
+            // Put the txs back in the heap
+            txs.extend(temp_vec.into_iter());
```

Comment:
Should this happen before or after `remove_all_txs` is checked?

Consider the following:
1. `max_retries` is 1000. Account A has nonce 0. A sends a transaction T1000 with nonce 1001.
2. 1000 blocks later, A sends a transaction T0 with nonce 0, and T2 with nonce 2.

What I would expect: T0 gets executed, while T2 remains with the queue, with `retries = 1`.
What seems will happen: T0 gets executed. T2 is popped and placed into `temp_vec` because it has 999 retries left. T1000 is checked, it now has 1000 retries, so it's dropped and `remove_all_txs` is set to true. Then, here, `temp_vec` (including T2) is poured back into `txs`, and in the if-statement just below `txs` gets cleared, thus T2 is dropped from the mempool.

Reply:
In my example there is no T1 though - T2 should remain in the queue for another 999 blocks (or until a T1 is broadcast later). The code seems like it will eject it immediately though, just because T1000 was ejected.

Reply:
Oh right, I misread, sorry about that. However, I still do not think it will happen as you say. 

There is a break whenever the TX nonce is too high, so what will happen is that only the lowest nonced TX has its retries incremented. That is to say, the flow would be: when T2 arrives, its retries will be incremented until == 1000, T1000 will not be touched and will stay on 999.

There is no point checking higher nonces, and implicitly if you have run out of retries for the lower one, the higher ones are forefit.

Reply:
Oh I see! Yeah I glossed over that `break` - I agree that it resolves my concern, and everything looks correct. Thanks for clearing it up for me.